### ISSUE TITLE
Support for Bottles Wine Prefixes

### ISSUE THREAD

**Issue by Heus-Sueh** [mtkennerly/ludusavi#268]
### What's your idea?

Support for prefixes created by Bottles which can look for wine prefixes in this directory:
`~/.var/app/com.usebottles.bottles/data/bottles/bottles`

**Comment by mtkennerly** [comment]
Hi! I'm not too familiar with Bottles. Is there some metadata that ties each prefix to a specific game, or is it an ad hoc/manual thing where you could install one or more games in a single prefix? If they're essentially normal Wine prefixes, then you could configure them in Ludusavi as a Wine prefix root with a glob pattern in the path. However, that would make the scan much slower, since it would have to scan every prefix for every game without some way to narrow it down.

**Comment by Heus-Sueh** [comment]
> Hi! I'm not too familiar with Bottles. Is there some metadata that ties each prefix to a specific game, or is it an ad hoc/manual thing where you could install one or more games in a single prefix? If they're essentially normal Wine prefixes, then you could configure them in Ludusavi as a Wine prefix root with a glob pattern in the path. However, that would make the scan much slower, since it would have to scan every prefix for every game without some way to narrow it down.

yes, you can install games in the same wineprefix, in this case the wine prefix that the user creates is in this directory: `/var/home/<user>/.var/app/com.usebottles.bottles/data/bottles/bottles`
![image](https://github.com/mtkennerly/ludusavi/assets/98482469/fa50037c-1b88-4078-b4b9-6fd0aea89877)

**Comment by VashStampedeTRI14** [comment]
Hope this helps but within this folder /home/<user>/.var/app/com.usebottles.bottles/data/bottles/ is a library.yml file that looks like below. This contains the list of all prefixes created within bottles.

61907121-a070-4297-a2d5-a350c7197314:
  bottle:
    name: Diablo IV
    path: Diablo-IV
  icon: /home/<user>/.var/app/com.usebottles.bottles/data/bottles/bottles/Diablo-IV/icons/Battle.net.png
  id: 2cd58ae9-15fd-41ab-b8c1-28af81cca37f
  name: Battle.net
  thumbnail: grid:b434b1ad-1377-4e56-9232-d021e9ab4148.png
d76a3e35-1d27-4e78-bfd2-ad39d2cda244:
  bottle:
    name: Grim Dawn
    path: Grim-Dawn
  icon: /home/<user>/.var/app/com.usebottles.bottles/data/bottles/bottles/Grim-Dawn/icons/Grim
    Dawn.png
  id: d907daf0-af13-45ac-85bc-3b10d512397e
  name: Grim Dawn
  thumbnail: grid:7e73de6d-fd7b-4b1d-ae51-7d7206dfe667.png



Then within each wine prefix is a bottles.yml which contains the .exe, name, ect. Example of this:
Arch: win64
CompatData: ''
Creation_Date: '2023-10-09 16:38:09.736543'
Custom_Path: false
DLL_Overrides: {}
DXVK: dxvk-2.3
Environment: Gaming
Environment_Variables: {}
External_Programs:
    d907daf0-af13-45ac-85bc-3b10d512397e:
        arguments: ''
        dxvk: true
        dxvk_nvapi: false
        executable: Grim Dawn.exe
        folder: /home/<user>/.var/app/com.usebottles.bottles/data/bottles/bottles/Grim-Dawn/drive_c/GOG
            Games/Grim Dawn
        fsr: false
        icon: com.usebottles.bottles-program
        id: d907daf0-af13-45ac-85bc-3b10d512397e
        name: Grim Dawn
        path: /home/<user>/.var/app/com.usebottles.bottles/data/bottles/bottles/Grim-Dawn/drive_c/GOG
            Games/Grim Dawn/Grim Dawn.exe
        pulseaudio_latency: false
        removed: null
        script: null
        virtual_desktop: false
        vkd3d: true
Installed_Dependencies:
- d3dx9
- msls31
- arial32
- times32
- courie32
- d3dcompiler_43
- d3dcompiler_47
- vcredist2010
- vcredist2012
- vcredist2013
- vcredist2015
Language: sys
LatencyFleX: latencyflex-v0.1.1
NVAPI: dxvk-nvapi-v0.6.4
Name: Grim Dawn
Parameters:
    custom_dpi: 96
    decorated: true
    discrete_gpu: true
    dxvk: true
    dxvk_nvapi: false
    fixme_logs: false
    fsr: false
    fsr_quality_mode: none
    fsr_sharpening_strength: 2
    fullscreen_capture: false
    gamemode: false
    gamescope: false
    gamescope_borderless: false
    gamescope_fps: 0
    gamescope_fps_no_focus: 0
    gamescope_fullscreen: true
    gamescope_game_height: 0
    gamescope_game_width: 0
    gamescope_scaling: false
    gamescope_window_height: 0
    gamescope_window_width: 0
    latencyflex: false
    mangohud: false
    mouse_warp: true
    obsvkc: false
    pulseaudio_latency: false
    renderer: gl
    sandbox: false
    sync: esync
    take_focus: false
    use_be_runtime: true
    use_eac_runtime: true
    use_runtime: false
    use_steam_runtime: false
    versioning_automatic: false
    versioning_compression: false
    versioning_exclusion_patterns: false
    virtual_desktop: false
    virtual_desktop_res: 1280x720
    vkbasalt: false
    vkd3d: true
    vmtouch: false
    vmtouch_cache_cwd: false
Path: Grim-Dawn
Runner: lutris-GE-Proton8-17-x86_64
RunnerPath: ''
Sandbox:
    share_net: false
    share_sound: false
State: 0
Uninstallers: {}
Update_Date: '2023-10-10 07:28:56.519460'
VKD3D: vkd3d-proton-2.10
Versioning: false
Versioning_Exclusion_Patterns: []
Windows: win7
WorkingDir: ''
data: {}
run_in_terminal: false
session_arguments: ''

**Comment by LordFransie** [comment]
I have this working manually.
You'd literally just need to modify the initializing script to scan for directories in ~/.var/app/com.usebottles.bottles/data/bottles/bottles or the flatpack directory.

**Comment by bpjoestar** [comment]
> I have this working manually. You'd literally just need to modify the initializing script to scan for directories in ~/.var/app/com.usebottles.bottles/data/bottles/bottles or the flatpack directory.

I love ludusavi and it'd be great if it searches in the bottles dir by default.
How did you change it manually? Do I just give Ludusavi the permission with Flatseal to search in  ~/.var/app/com.usebottles.bottles/data/bottles/bottles?

**Comment by mtkennerly** [comment]
@bpjoestar It would *mostly* work if you both:

* Use Flatseal to give Ludusavi access to `~/.var/app/com.usebottles.bottles/data/bottles`
* Add a Wine root with a glob in Ludusavi: `~/.var/app/com.usebottles.bottles/data/bottles/bottles/*`

However, it won't correctly resolve `<base>` placeholders from the manifest because it won't know where each game is installed inside of a given prefix. It should still find a decent amount of save data, but not everything.

I'm still interested in adding proper support for this, but haven't quite figured out the best way to implement it, especially since Ludusavi would need to support both one-game-per-bottle and many-games-per-bottle. I almost think that Wine/Bottles roots would need to support a configurable list of sub-roots (e.g., one bottle could contain a GOG sub-root `drive_c/GOG
Games` and a Steam sub-root `drive_c/Program Files (x86)/Steam`).
